lsp/src/main/kotlin/org/gradle/declarative/lsp/DeclarativeTextDocumentService.kt

```diff
@@ -345,14 +357,70 @@ private fun computePropertyCompletions(
+    return dataClass.properties.mapNotNull { property ->
+        when (val resolvedType = SchemaTypeRefContext(analysisSchema).resolveRef(property.valueType)) {
+            is EnumClass -> completionItem(property, resolvedType)
+            is DataType.BooleanDataType -> completionItem(property, resolvedType)
+            is DataType.IntDataType -> completionItem(property, resolvedType)
+            is DataType.LongDataType -> completionItem(property, resolvedType)
+            is DataType.StringDataType -> completionItem(property, resolvedType)
+            else -> null
+        }
+    }
+}
 
-        CompletionItem("$propertyName = $targetType").apply {
-            kind = CompletionItemKind.Field
-            insertTextFormat = InsertTextFormat.Snippet
-            insertText = "${property.name} = ${computeTypedPlaceholder(1, property.valueType, analysisSchema)}"
+private fun completionItem(property: DataProperty, resolvedType: DataType) =
+    CompletionItem("${property.name} = ${property.valueType.toSimpleName()}").apply {
+        kind = CompletionItemKind.Field
+        insertTextFormat = InsertTextFormat.Snippet
+        insertText = "${property.name} = ${computeTypedPlaceholder(1, resolvedType)}"
+    }
+
+private typealias LabelAndInsertText = Pair<String, String>
+
+private fun computePropertyByValueFactoryCompletions(
+    dataClass: DataClass,
+    analysisSchema: AnalysisSchema
+): List<CompletionItem> {
+    fun indexValueFactories(analysisSchema: AnalysisSchema, type: DataClass, namePrefix: String): Map<FqName, List<LabelAndInsertText>> {
+        val factoryIndex = mutableMapOf<FqName, List<LabelAndInsertText>>()
+        type.memberFunctions
+            .filter { it.semantics is FunctionSemantics.Pure && it.returnValueType is DataTypeRef.Name }
+            .forEach {
+                val indexKey = (it.returnValueType as DataTypeRef.Name).fqName
+                val labelAndInsertText = "$namePrefix${computeCompletionLabel(it)}" to "$namePrefix${computeCompletionInsertText(it, analysisSchema)}"
+                factoryIndex.merge(indexKey, listOf(labelAndInsertText)) { oldVal, newVal -> oldVal + newVal }
+            }
+        type.properties.filter { it.valueType is DataTypeRef.Name }.forEach {
+            when (val propType = analysisSchema.dataClassTypesByFqName[(it.valueType as DataTypeRef.Name).fqName]) {
+                is DataClass -> {
+                    val propName = it.name
+                    val propIndex = indexValueFactories(analysisSchema, propType, "$namePrefix${propName}.")
+                    propIndex.forEach { (key, value) -> factoryIndex.merge(key, value) { oldVal, newVal -> oldVal + newVal } }
+                }
+
+                is EnumClass -> Unit
+                null -> Unit
+            }
+        }
+        return factoryIndex
+    }
+
+    val factories = indexValueFactories(analysisSchema, analysisSchema.topLevelReceiverType, "")
```

Comment:
💭 Currently, in addition to the top-level type, you can use the members of the nested containers. For instance, in 

```kotlin
foo {
    bar { 
        baz = qux.quux()
    }
}
```

`qux` could be a member of:
* `bar`'s receiver
* `foo`'s receiver
* the top-level receiver

We should probably discuss if we want to keep it that way, but if we do, these nested receivers should also contribute the value factory groups to completion.

Reply:
As it works now, `qux` will be found regardless which of those receivers it's on and will be proposed as a potential value for properties with the same type as the return type of the `quux()` function.

I think the real problem we should think about is that `qux.quux()` will be proposed as a potential value for `baz` even if `qux` comes from a totally unrelated block, which is part of the top-level receiver, but is unrelated to `foo` (thus by extension unrelated to `bar` and `baz` too).

Reply:
Added a TODO.